File: J9Ch/src/J_7_Two_Pointers/Optional_11/_382_Triangle_Count_Medium.java
package J_7_Two_Pointers.Optional_11;
import java.util.*;
/** 382 Triangle Count


 * Created by tianhuizhu on 6/28/17.
 */
public class _382_Triangle_Count_Medium {

    public class Solution {
        /**
         * @param S: A list of integers
         * @return: An integer
         */
        public int triangleCount(int S[]) {
            // write your code here
            int left = 0, right = S.length - 1;
            int ans = 0;
            Arrays.sort(S);
            for(int i = 0; i < S.length; i++) {
                left = 0;
                right = i - 1;
                while(left < right) {
                    if(S[left] + S[right] > S[i]) {
                        ans = ans + (right - left);
                        right --;
                    } else {
                        left ++;
                    }
                }
            }
            return ans;
        }
    }

}

File: J9Ch/src/S_1_Follow_up_in_CodeInterview/Optional_6/python_cpp/_382_Triangle_Count_Medium.cpp



class Solution {
public:
    int triangleCount(vector<int> &S) {
        sort(S.begin(), S.end());
        int ans = 0;
        for(int i = 0; i < S.size(); i++) {
            int k = i + 2;
            for(int j = i + 1; j < S.size(); j++) {
                while(k < S.size() && S[i] + S[j] > S[k])
                    k++;
                ans += k - j - 1;
            }

        }
        return ans;
    }
};